### DISCUSSION TITLE
How to get velocity vector(NED)  from GPS NMEA data?

### DISCUSSION THREAD

**Discussion by scott198510** [rodralez/NaveGo#98] [Ideas]
Mr Doc,
The GNSS speed in your code is obtained directly. As we all know, the original GNSS speed value is only one，such as NMEA format. 
Then how to use the original GNSS data which is NMEA format to get the velocity  in three directions?

Anyway, at present, the published papers or books directly obtain three values for GNSS velocity, but there is only one value in original GNSS data. Why not explain this? Can we say that the default is to use the data which is dfferential data(such as RTK)?

https://stackoverflow.com/questions/55554306/how-to-get-ned-velocity-from-gps

http://aprs.gids.nl/nmea/

**Comment by wangfeng35** [comment]
IN NEMA xxVTG and xxRMC, the GNSS speed is the 2D Speed Over Ground ignoring the altitude.  In order to get the 3D velocity components, you need to use the GNSS receiver that supports to output the 3D velocity in its own protocol.  For example, here: https://portal.u-blox.com/s/question/0D52p00008HKDe8CAH/2d-or-3d-gps-speed .

**Reply by scott198510** [reply]
@wangfeng35 If you get the data such as NEMA format , no gnss receiver, how can you do that?
As we all know, theoretically, the speed given by GNSS receiver can output the velocity values in three directions, but it is not always given in practice.

Assuming that you have only one data file output in NEMA format, how do you use Kalman filter algorithm to solve the integrated navigation problem.

In NaveGo, the  algorithm get 3d gnss velocity directly, how can do that?


![image](https://user-images.githubusercontent.com/17826746/154950204-4112974e-54d7-4de8-88ef-3281c27a6180.png)

**Reply by wangfeng35** [reply]
As for the north and east velocity, use the speed over ground and course over ground (sin/cos). For the up velocity, use the adjacent altitudes in the GGA divided by the time interval to estimate. Notice, the speed in GPS is unreliable when the speed is low.

**Reply by scott198510** [reply]
@wangfeng35 If it exists three  Euler angles， you only consider  speed over ground and course over ground, then you use math trigonometric function （sin/cos）, you can not get true velocity?

**Reply by wangfeng35** [reply]
The NMEA message does not include the three Euler angles, and the speed over ground  in NMEA is only 2D speed.

**Reply by scott198510** [reply]
@wangfeng35 Ok！！ The answer of yours is so instant. 

In fact, I have considered this method earlier. For the two velocities in the horizontal plane, we adopt the decomposition method, for the vertical one, we adopt the altitude data difference. But in fact, I want to reproduce the three velocities of GNSS receiver, any other methods?

**Reply by wangfeng35** [reply]
No other methods occur to me right now besides using the existing information in the NMEA message. 
The best way is to use the GNSS receiver that supports the output of three velocity, which is a bit difficult. 

Besides, you can also use the position-only update with GPS position vector (lat/lon/alt) in NaveGo. In this way, you need to modify the measurement vector to GPS position vector and also H matrix. However, the performance may be degraded.

**Reply by scott198510** [reply]
@wangfeng35  From what you mean, you seem to have used NMEA data for integrated navigation solution, right？
In fact, our supplier here has achieved good solution by using the NMEA data which provided by us, so I'm just curious about how they use the GNSS data. Is it really like the three rate values obtained by the above method? Maybe...

**Reply by scott198510** [reply]
Anyway, in fact, there are NMEA format data that are either solved or used, but most of the literatures/papers/books do not mention it. By default, the receiver is used to directly obtain three effective velocity values. Right?

**Reply by rodralez** [reply]
> In fact, I have considered this method earlier. For the two velocities in the horizontal plane, we adopt the decomposition method, and the the vertical one, we adopt the altitude data difference. But in fact, I want to reproduce the three velocities of GNSS receiver, any other methods?

The GNSS sensor estimates positions and velocities using different methods. Velocities are obtained by Doppler shifts or carrier phase measurements [1]. This is beneficial since GNSS positions and velocities tend to be uncorrelated, one of the core assumptions in Kalman filter implementation.

When Vn and Ve are estimated from GNSS speed-over-ground velocity and GNSS course-over-ground angle, I am not sure if these estimations are uncorrelated from positions. On the other hand, Vd obtained from the derivative of altitude will be correlated to altitude for sure. So, this is a sub-optimal approach.

In summary, the best approach is to get a GNSS sensor that directly delivers Vn, Ve and Vd.

[1] https://insidegnss.com/wp-content/uploads/2018/01/marapr15-SOLUTIONS.pdf

**Reply by rodralez** [reply]
> Besides, you can also use the position-only update with GPS position vector (lat/lon/alt) in NaveGo. In this way, you need to modify the measurement vector to GPS position vector and also H matrix. However, the performance may be degraded.

Yes, according to my experience, a loosely-coupled INS/GNSS system with position-only updates is useless, at least using low-cost IMU sensors.

**Reply by wangfeng35** [reply]
> yway, in fact, there are NMEA format data that are either solved or used, but most of the literatures/papers/books do not mention it. By default, the receiver is used to directly obtain three effective velocity values. Right?

Yes, if the velocity update is used, the receiver they used can directly output the three velocity components.

**Reply by wangfeng35** [reply]
> > Besides, you can also use the position-only update with GPS position vector (lat/lon/alt) in NaveGo. In this way, you need to modify the measurement vector to GPS position vector and also H matrix. However, the performance may be degraded.
> 
> Yes, according to my experience, a loosely-coupled INS/GNSS system with position-only updates is useless, at least using low-cost IMU sensors.

You're right. The best way is to use both position and velocity update when a low-cost IMU is used.
